### ISSUE TITLE
let_value decay-copies its final result

### ISSUE THREAD

**Issue by ispeters** [facebookexperimental/libunifex#556]
See discussion on #545 where I discovered the problem. The successor receiver in `let_value` tries to protect itself against the successor invoking `set_value` (and, incidentally, `set_error`) with a reference to a value that lives in the successor operation state by decay-copying `set_value`'s arguments onto the stack, which breaks `let_value`'s contract with its receiver if `let_value` has advertised that it's going to produce lvalue references; if we're going to distinguish between producing rvalue references and producing values, it breaks the contract when we've promised rvalue references, too, since the result object identities change.

Copying from my comment on #545, this test in `finally_test.cpp` repros the build error:
```diff
+
+TEST(Finally, CombinedWithLetValue) {
+  const int i{42};
+  auto ret = let_value(
+                 just(&i),
+                 [](const int* pi) noexcept {
+                   return just(pi) |
+                       then([](const int* pi) -> const int& { return *pi; });
+                 }) |
+      finally(just()) | then([](const int& i) { return &i; }) | sync_wait();
+
+  ASSERT_TRUE(ret.has_value());
+  EXPECT_EQ(&i, *ret);
+}
```
The problem is [this](https://github.com/facebookexperimental/libunifex/blob/main/include/unifex/let_value.hpp#L65) decay-copy in `let_value`:
```cpp
  template <typename... SuccessorValues>
  void set_value(SuccessorValues&&... values) && noexcept {
    auto& op = op_;
    UNIFEX_TRY {
      // Taking by value here to force a copy on the offchance the value
      // objects lives in the operation state (e.g., just), in which
      // case the call to cleanup() would invalidate them.
      [&](auto... copies) {
        cleanup();
        unifex::set_value(
            std::move(op.receiver_), (decltype(copies) &&) copies...);
      } ((SuccessorValues&&) values...);
    } UNIFEX_CATCH (...) {
      unifex::set_error(std::move(op.receiver_), std::current_exception());
    }
  }
```

The `let_value` in the new test case is advertising that it produces a `const int&` but then it produces `int`/`int&&`.

**Comment by ispeters** [comment]
I asked whether the stdexec implementation of `let_value` and `finally` have this unfortunate interaction and the answer was no, because, apparently, their `let_value` doesn't clean up the successor operation before invoking `set_value` so there's no need to worry about receiving a reference to a subobject of the successor operation state.

I'm a little bit worried about deferring destruction of the successor operation to the destructor of the `let_value` operation, but it's really just a vague anxiety.  It's probably worth trying.  I'll fool around with the problem today and see what I can learn.

**Comment by ispeters** [comment]
I have a local change that makes the above-described unit test build and pass, but I'm finding more problems elsewhere that I haven't minimized, yet.  I'll publish a `let_value` PR today, but we're going to have to track down all the lvalue reference problems to set this straight, I think.

**Comment by ispeters** [comment]
This particular issue is fixed with PR #557. We'll have to file follow-up issues for the other algorithms that have similar bugs.
